converting more components to setup composition api

aweebs opened the pull request:
Converts the remaining components in the `vite` branch to setup composition api.

Replaces v-data-table usage in apikeys and devices with regular tables, since the data-table component is not available yet but also is likely overkill for this purpose. Let me know if you'd like these reverted.

Recommend reviewing commit by commit and **hiding whitespace changes**. 

### TODO
- [x] test logs and activities page
- [x] fix activity and log layouts
- [x] test devices page
- [x] remove SelectedDeviceInfo component since all data is in the table already
- [x] test ImageEditor and ImageSearch
- [x] test MetadataEditor, MetadataEditorDialog, and PersonEditor
- [x] test updated route pages
- [x] test setup wizard - couldn't fully test until vuetify 3 stepper is released
- [x] Change `useDataFns` to return direct value rather than ref for now because of https://github.com/vuejs/core/pull/7306

### Follow up work (for future PRs)
- Connect `UpNext` button again
- metadata image adding

aweebs commented on frontend/src/pages/settings/logs-and-activity.vue:
```diff
@@ -115,14 +111,14 @@
   </settings-page>
 </template>
 
-<route lang="yaml">
-meta:
-  admin: true
-</route>
```
I removed this since it wasn't letting my admin account access the page. I didn't see it on the api keys or devices page which I think are admin only as well. Should these all be removed or is there a more correct way to perform this check?

ferferga reviewed (changes requested):
First round of feedback, will carefully review the files later:

The layout at logs page is a bit borked (in master, the arrow is at the left):
![image](https://user-images.githubusercontent.com/10274099/222081344-55b8fcc6-fab0-477b-bd9d-0a7f3f6713ca.png)

I don't remember why we had the **SelectedDeviceInfo** component in the first place, the info it provides is the same that can be located at the table. @ThibaultNocchi do you remember something?
I think we can get off without it. In case we go for it again, the table should have a hover effect, so it's obvious for the user that there's an action on click.

The data returned by useDataFns has double quotes. This is because it's a ref, and it's handled upstream at https://github.com/vuejs/core/pull/7306. For now, I would say just unwrap ``.value`` there.

In any case, those functions could be moved out of the template imo.

The check for the admin pages lives at ``plugins/router/middlewares/admin-pages.ts``. If you look at the check, the redirection is made when the user is an admin, which should be changed to when the user IS NOT an admin.

aweebs commented on frontend/src/components/Playback/UpNext.vue:
```diff
@@ -67,75 +67,57 @@
 
-<script lang="ts">
-import { defineComponent } from 'vue';
+<script setup lang="ts">
```
@ferferga @ThibaultNocchi I didn't see this component referenced anywhere. Is this something that will be reconnected at some point or can it be discarded?

aweebs replied:
Adding this to my list of future TODOs, but excluding it from the scope of this PR

aweebs commented:
@ferferga this is ready for a proper review, I've tested the changes locally fairly well to confirm things appear correct.

ThibaultNocchi commented on frontend/src/components/Item/RelatedItems.vue:
```diff
@@ -1,49 +1,47 @@
+    <!-- <v-skeleton-loader v-else-if="loading" type="heading" /> -->
+    <v-list bg-color="transparent" lines="two">
+      <div v-if="!loading && relatedItems.length > 0">
+        <v-list-item
+          v-for="relatedItem in relatedItems"
+          :key="relatedItem.Id"
+          :to="getItemDetailsLink(relatedItem)">
```
```suggestion
        <v-list-item
          v-for="relatedItem in relatedItems"
          :key="relatedItem.Id"
          :to="getItemDetailsLink(relatedItem)"
          :title="relatedItem.Name"
          :subtitle="relatedItem.ProductionYear">
```

to free up the HTML below

ThibaultNocchi commented on frontend/src/components/Playback/UpNext.vue:
```diff
@@ -67,75 +67,57 @@
-  methods: {
-    getEndsAtTime,
-    getRuntimeTime
+const currentItemDuration = computed((): number => {
+  return ticksToMs(playbackManager.currentItem?.RunTimeTicks) / 1000;
+});
+const currentItemTimeLeft = computed((): number => {
+  return Math.round(
+    currentItemDuration.value - (playbackManager.currentTime || 0)
+  );
+});
```
I'd drop the returns

ThibaultNocchi commented on frontend/src/components/Playback/UpNext.vue:
```diff
@@ -67,75 +67,57 @@
+
+  return 45;
 });
+const visible = computed(
+  (): boolean | undefined =>
+    !isHiddenByUser.value &&
+    playbackManager.currentlyPlayingMediaType === 'Video' &&
+    currentItemTimeLeft.value <= nextUpDuration.value
+);
```
You can drop undefined and maybe the whole type I guess (depends on the linter :p)

ThibaultNocchi commented on frontend/src/pages/item/_itemId/index.vue:
```diff
@@ -246,94 +246,62 @@ import { getItemDetailsLink, getMediaStreams } from '@/utils/items';
+const currentAudioTrack = ref<number>();
+const currentSubtitleTrack = ref<number>();
+
+const crew = computed((): BaseItemPerson[] => {
+  if (!item.value?.People) {
+    return [];
+  }
+
+  return item.value.People.filter((person: BaseItemPerson) =>
+    ['Director', 'Writer'].includes(person?.Type ?? '')
+  );
+});
```
I think this can be written as

```suggestion
const crew = computed(
  (): BaseItemPerson[] =>
    item.value?.People?.filter((person: BaseItemPerson) =>
      ['Director', 'Writer'].includes(person?.Type ?? '')
    ) ?? []
);
```

ThibaultNocchi commented on frontend/src/pages/item/_itemId/index.vue:
```diff
@@ -246,94 +246,62 @@ import { getItemDetailsLink, getMediaStreams } from '@/utils/items';
+  );
+});
+
+const actors = computed((): BaseItemPerson[] => {
+  if (!item.value?.People) {
+    return [];
+  }
+
+  return item.value.People.filter(
+    (person: BaseItemPerson) => person.Type === 'Actor'
+  ).slice(0, 10);
+});
```
Same as before?

ThibaultNocchi commented on frontend/src/pages/person/_itemId/index.vue:
```diff
@@ -125,98 +126,109 @@ import { getBlurhash } from '@/utils/images';
-  item.PremiereDate
-    ? useDateFns(format, new Date(item.PremiereDate), 'PPP').value
+  item.value.PremiereDate
+    ? useDateFns(format, new Date(item.value.PremiereDate), 'PPP').value
     : undefined
 );
 
 const deathDate = computed(() =>
-  item.EndDate
-    ? useDateFns(format, new Date(item.EndDate), 'PPP').value
+  item.value.EndDate
+    ? useDateFns(format, new Date(item.value.EndDate), 'PPP').value
     : undefined
 );
```
It seems those values are now undefined

aweebs replied:
It broke from the DateFns change, I've reverted that and applied the fix specifically in the components consuming it, which has also fixed these not displaying. I'm surprised this didn't cause a type error thought :S

ThibaultNocchi commented on frontend/src/pages/series/_itemId/index.vue:
```diff
@@ -147,75 +147,52 @@ import { getBlurhash } from '@/utils/images';
-    const item = (
+const item = ref<BaseItemDto>({});
+
+const crew = computed((): BaseItemPerson[] => {
+  if (!item.value?.People) {
+    return [];
+  }
+
+  return item.value.People.filter((person: BaseItemPerson) =>
+    ['Director', 'Writer'].includes(person?.Type ?? '')
+  );
+});
+
+const actors = computed((): BaseItemPerson[] => {
+  if (!item.value?.People) {
+    return [];
+  }
+
+  return item.value.People.filter(
+    (person: BaseItemPerson) => person.Type === 'Actor'
+  ).slice(0, 10);
+});
```
Same changes as before

ferferga reviewed (changes requested):
> Change useDataFns to return direct value rather than ref for now

When I started the migration to Vue 3, I approached this problem like you did here. However, this is not correct because of the following reasons (in fact the useDateFns has been recently modified to always return refs):

* I recall returning values directly broke reactivity in some situations, so it was inconsistent. Reactive effects (Ref or ComputedRef) always ensured a consistent experience in that regard.
* Because you're returning a value, I believe the effects are not disposed properly when the component is unmounted, which could leak to memory leaks.
* You don't really know if something is reactive or not when using primitives, but you do when using effects/refs. DX-wise this is much better when you know when stuff is reactive or not.

I know it's painful, but hopefully this gets fixed out soon. I'm already tracking that issue because I myself reported it in the Vue discord and fixing this in the future it's a matter of making a PR once the fix is upstream to replace ``.value }}`` with `` }}``

ThibaultNocchi commented on frontend/src/components/Item/CollectionTabs.vue:
```diff
@@ -41,10 +44,12 @@ watch(
   () => props.item,
   async (item) => {
     if (!children.value) {
-      loading.value = true;
-      await items.fetchAndAddCollection(item.Id);
-      loading.value = false;
```
I agree with that sentiment, but I think you just forgot to change `!children.value` into `children.value` :)

aweebs replied:
Well that's certainly a whoosh moment haha, fix incoming!

aweebs replied:
I've also added a comment here, since this isn't standard loading behaviour from other parts of the app

ThibaultNocchi commented on frontend/src/components/Item/CollectionTabs.vue:
```diff
@@ -40,11 +43,14 @@ const children = computed(() => {
-      loading.value = true;
-      await items.fetchAndAddCollection(item.Id);
-      loading.value = false;
+    // only try to load children if there aren't any children items in the itemStore yet
```
```suggestion
    /**
    * Only try to load children if there aren't any children items in the itemStore yet
    */
```

We tend to prefer this syntax :)

aweebs replied:
I'll update this for consistency, but I'm curious why, I'll put some thoughts in Matrix.

ThibaultNocchi commented on frontend/src/components/Item/Metadata/ImageSearch.vue:
```diff
@@ -3,32 +3,36 @@
           variant="outlined"
-          hide-details />
+          hide-details
+          clearable />
```
This sets the value to `null` and makes the next queries fetch a `null` provider, instead of removing the parameter. I think there should a check for null in the query below :)

ThibaultNocchi commented on frontend/src/components/Item/CollectionTabs.vue:
```diff
@@ -1,21 +1,24 @@
+  <div v-if="children">
     <v-tabs v-model="currentTab" class="mb-3" bg-color="transparent">
-      <v-tab v-for="(items, type) in children" :key="type">
+      <v-tab v-for="(items, type) in children" :key="type" :value="type">
         {{ type }} ({{ items.length }})
       </v-tab>
```
```suggestion
      <v-tab v-for="(baseItems, type) in children" :key="type" :value="type">
        {{ type }} ({{ baseItems.length }})
      </v-tab>
```

ThibaultNocchi replied:
To avoid conflict with the items store and the linter being angry

ThibaultNocchi commented on frontend/src/components/Item/CollectionTabs.vue:
```diff
@@ -1,21 +1,24 @@
-      <v-tab v-for="(items, type) in children" :key="type">
+    <v-window v-model="currentTab" class="bg-transparent">
+      <v-window-item
+        v-for="(items, type) in children"
+        :key="type"
+        :value="type">
         <v-container>
           <skeleton-item-grid v-if="loading" :view-type="item.Type" />
           <item-grid :loading="loading" :items="items" />
```
```suggestion
        v-for="(baseItems, type) in children"
        :key="type"
        :value="type">
        <v-container>
          <skeleton-item-grid v-if="loading" :view-type="item.Type" />
          <item-grid :loading="loading" :items="baseItems" />
```

aweebs commented:
@ferferga @ThibaultNocchi this is ready to go if you don't have any further change requests

ferferga commented on frontend/src/pages/wizard.vue:
```diff
@@ -78,82 +78,73 @@ meta:
+const wizardStage = ref(1);
+const maxWizardStage = ref(1);
+
+const heading = computed((): string => {
```
Probable unnecessary type here

ferferga reviewed (changes requested):
Thank you very much! This is massive, I'm really grateful for this.

Sorry for the huge amount of comments. Most of them are nitpicks, some will require more work. Hopefully my "strictness" (before I didn't care as much in reviews, but with the Vue 3 migration I discovered the amount of legacy we would have saved us nowadays if we had been a bit more strict in reviews. I don't want that anymore) doesn't turn you down! :)

aweebs commented on frontend/src/components/Buttons/PlayButton.vue:
```diff
@@ -48,12 +48,20 @@ const props = withDefaults(
+    fab: false,
+    shuffle: false,
+    disabled: false,
+    videoTrackIndex: undefined,
+    audioTrackIndex: undefined,
+    subtitleTrackIndex: undefined
```
When I don't provide these  I get a linting error `Prop 'subtitleTrackIndex' requires default value to be set.eslint[vue/require-default-prop](https://eslint.vuejs.org/rules/require-default-prop.html)`

aweebs commented on frontend/src/components/Item/ItemMenu.vue:
```diff
@@ -33,7 +33,7 @@
     </v-menu>
     <metadata-editor-dialog
-      v-if="metadataDialog"
+      v-if="item.Id"
```
`v-if` evaluations won't get added to the DOM (at least to my knowledge and I quickly checked the DOM).
This `v-if` directive is here because the MetadataEditorDialog component always expects a defined `item-id` parameter, and because item.Id could be null or undefined, adding the `v-if` narrows the type to remove null or undefined and satisfies the type checking. In practice it will always be true and is only there to satisfy typings.

ThibaultNocchi replied:
Also, almost nothing is added to the DOM by just adding metadata-editor-dialog, as it's a v-dialog the content is also v-ifed by Vuetify

ferferga replied:
@ThibaultNocchi @aweebs This was not the case with Vuetify 2. Are we sure that in Vuetify 3 everything is v-ified? Can any of you check please? I'm afk right now.

In case everything is v-ified now, forget about this.

aweebs replied:
Confirmed that nothing in the component is mounted into the DOM when the dialog is not opened.

aweebs commented on frontend/src/components/Item/Metadata/ImageEditor.vue:
```diff
@@ -10,7 +10,7 @@
             variant="outlined"
             width="288">
-            <v-img :src="imageFormat(item)" contain width="288" height="162" />
+            <v-img :src="imageFormat(item)" :width="288" :aspect-ratio="1.78" />
```
I removed both `width` and `contain` because they're not defined on the v-img api (https://vuetifyjs.com/en/api/v-img/) and were throwing type errors. I know the vuetify 3 docs are lacking at the moment and that may just be an error.

The added aspect ratio is to achieve the same dimensions as before (288 / 1.78 ~= 162, which is the old height).

So the display of these images should be the same and be responsive as they were before.

aweebs replied:
That being said, I'll look to putting these into some responsive classes to make them reflow nicely. I'll use width 100% and the existing aspect ratio.

aweebs replied:
Ok, so these are updated to use vuetify's breakpoints to display a different number of images based on the screen size in both ImageEditor and ImageSearch. The aspect ratio that is used to display different types of images is now pulled from a common utility function too. Let me know what you think, happy to tweak this a bit more.

aweebs commented on frontend/src/components/Item/Metadata/ImageSearch.vue:
```diff
@@ -90,179 +90,174 @@
+const images = ref<RemoteImageInfo[]>([]);
+const loading = ref(true);
+
+const types = computed((): { text: string; value: ImageType }[] => [
```
I left this in because TypeScript will infer the type as union of the different values in this list (which are a set of unique strings). Which means that theoretically someone could add an entry to this list which has a value that isn't a valid `ImageType`. Including this type prevents that since there will an error if that happens. Keeping types as narrow as possible is good practice so I've added it in.

aweebs commented on frontend/src/components/Item/Metadata/ImageSearch.vue:
```diff
@@ -90,179 +90,174 @@
-        throw new Error('Expected image type and url to be present');
-      }
+const ratio = computed(() =>
+  type.value === ImageType.Backdrop ? '1.777777778' : '0.666666667'
```
These are just hardcoded aspect ratios so that all of the image cards have the same height when being displayed.

aweebs commented on frontend/src/components/Item/Metadata/MetadataEditor.vue:
```diff
@@ -1,163 +1,187 @@
+          <v-list lines="two">
+            <v-list-item
+              :title="t('addNewPerson')"
+              @click="handlePersonEdit({})">
```
I'll add the intermediate function. I'm going to leave the name since I think "handlePersonEdit" is a bit more explicit.

aweebs commented on frontend/src/components/Item/Metadata/PersonEditor.vue:
```diff
@@ -46,73 +51,59 @@
         }">
         <v-spacer />
         <v-btn variant="flat" width="8em" class="mr-1" @click="handleCancel">
```
Change to an immediate emit!
Do you mean rename the emitted events to use `on` prefix like `onClose`? It doesn't seem we do that anywhere else, is that a Vue naming convention?

aweebs commented on frontend/src/components/Item/Metadata/PersonEditor.vue:
```diff
@@ -46,73 +51,59 @@
-    }
+const editState = ref<BaseItemPerson>();
+const options = computed(() => [
+  { text: t('actor'), value: 'Actor' },
+  { text: t('composer'), value: 'Composer' },
+  { text: t('director'), value: 'Director' },
+  { text: t('guestStar'), value: 'GuestStar' },
+  { text: t('producer'), value: 'Producer' },
+  { text: t('writer'), value: 'Writer' }
```
These adjust the `Type` field on `BaseItemPerson`, which is defined as `string`. Form the Jellyfin server code, it's [defined as a string](https://github.com/jellyfin/jellyfin/blob/f3c57e6a0ae015dc51cf548a0380d1bed33959c2/MediaBrowser.Model/Dto/BaseItemPerson.cs#L36). However, there is a [PersonType.cs](https://github.com/jellyfin/jellyfin/blob/f3c57e6a0ae015dc51cf548a0380d1bed33959c2/MediaBrowser.Model/Entities/PersonType.cs) class which seems to be an enumeration of all the possible values, but for some reason it's not an enum (might be a fun first bug for working on the server code :D). Since it's not an enum, the TS SDK types aren't picked up for it though.

TL;DR It's not a `BaseItemKind` and we don't have complete types for it :(

ferferga replied:
BaseItemKind didn't exist before either, it was added by @crobibero. There's something wonky in those enums that it's not caught up by the OpenAPI server generator.

Maybe he can give you some insights in case you want to dive in server land :P

crobibero replied:
I looked at making PersonType an enum but it can be complete customized by plugins.. so while it is possible we would need to also migrate the database and make a lot of server code changes

aweebs replied:
Oof that's unfortunate that's it could be anything :( I'll leave this as is for now

aweebs commented on frontend/src/components/Item/RelatedItems.vue:
```diff
@@ -71,36 +67,37 @@ const props = withDefaults(
 
 const relatedItems = ref<BaseItemDto[]>([]);
-const loading = ref(true);
+const loading = ref(false);
```
It doesn't matter in practice, since the loading state will immediately be flipped to on, but just for consistency. All the other components with a loading property start with this `false` (actually I messed up ImageSearch when I re-wrote it, I'll update that too though.

I don't have an opinion for whether defaulting true or false is more correct, so I'm also happy to update all components using a `loading` variable to initialize to `true` if you prefer that more.

ferferga replied:
@aweebs I don't recall any place where it's used with it initialized to false, but that doesn't make much sense imo, so ofc I'm all in for changing the defaults to true and avoid the extra assignment (and possibly DOM re-renders)

aweebs replied:
Went through and you're correct, the others uses should be false to start off. With this one it should be true since the first mount action should always be a load (unless the prop item.Id isn't defined on first load, but that's probably enough of an edge case not to worry about). I'll update this!

aweebs commented on frontend/src/pages/artist/_itemId/index.vue:
```diff
@@ -123,31 +119,43 @@ import {
+const albums = ref<BaseItemDto[]>([]);
+const appearances = ref<BaseItemDto[]>([]);
+const musicVideos = ref<BaseItemDto[]>([]);
+const activeTab = ref(3);
```
Done haha, I agree, that was a little strange!

aweebs commented on frontend/src/pages/artist/_itemId/index.vue:
```diff
@@ -123,31 +119,43 @@ import {
+);
 
-    const item = (
+watch(
```
I don't think this can be written as a computed. This function does a bunch of state assignment which is a side effect and is [not supposed to happen in computed functions](https://vuejs.org/guide/essentials/computed.html#getters-should-be-side-effect-free).

aweebs commented on frontend/src/pages/series/_itemId/index.vue:
```diff
@@ -147,75 +147,44 @@ import { getBlurhash } from '@/utils/images';
+);
+
+const writers = computed(() =>
+  crew.value.filter((person) => person.Type === 'Writer')
```
See above comment, but sadly no. These are different enums but the TypeScript SDK doesn't have them :(

aweebs commented on frontend/src/pages/item/_itemId/index.vue:
```diff
@@ -246,94 +246,54 @@ import { getItemDetailsLink, getMediaStreams } from '@/utils/items';
+const currentAudioTrack = ref<number>();
+const currentSubtitleTrack = ref<number>();
+
+const crew = computed((): BaseItemPerson[] =>
+  (item.value?.People ?? []).filter((person) =>
+    ['Director', 'Writer'].includes(person?.Type ?? '')
+  )
+);
```
I left this here and in the other spots in this file and a few others because the return value reported by TypeScript and VS Code intellisense was the literal representation of the value (eg `{ Id: string | null | undefined, Name: string | null | undefined, ... }` instead of just `BaseItemPerson`). This just makes intellisense show the return value type as `BaseItemPerson`, so its purely for developer convenience.

Separately, I've updated all computed values that specify a return value to use the `computed<T>` form, I actually prefer that more too!

ferferga replied:
@aweebs Fair enough, thanks!

aweebs commented on frontend/src/pages/settings/apikeys.vue:
```diff
@@ -4,130 +4,184 @@
-            </p>
-          </template>
-        </v-data-table> -->
+        <v-table>
```
I think that's just your view on github, there aren't any blank lines in the template portion of this file.

aweebs commented on frontend/src/pages/artist/_itemId/index.vue:
```diff
@@ -160,25 +168,27 @@ export default defineComponent({
     );
 
-    const appearances =
+    appearances.value =
```
Ah yeah, right, the 3 values before this can just be computed. Good catch!

aweebs commented on frontend/src/pages/metadata/index.vue:
```diff
@@ -17,88 +17,75 @@
-  }
-});
+/** set selected itemId */
+function handleAction(ids: string[]): void {
```
Went with `onExpandItem` :)

aweebs commented on frontend/src/pages/settings/logs-and-activity.vue:
```diff
@@ -1,112 +1,103 @@
+            <v-list-item
+              v-for="file in logs.data"
+              :key="file.Name ?? undefined"
+              :href="getLogFileLink(file.Name ?? '')"
```
The html helpers only have html sanitization, this a url. SonarCloud doesn't seem to have any issues, and I think that this call is fine especially since the user supplied data is templated as a query param so there's no sanitization needed in this case.

aweebs commented on frontend/src/pages/settings/logs-and-activity.vue:
```diff
@@ -136,98 +132,128 @@ import IMdiStop from 'virtual:icons/mdi/stop';
+function getColorFromSeverity(severity: LogLevel | undefined): string {
+  switch (severity) {
+    case LogLevel.Trace: {
+      return theme.current.value.colors.success;
+    }
+    case LogLevel.Debug: {
+      return theme.current.value.colors.accent;
+    }
+    case LogLevel.Information: {
+      return theme.current.value.colors.info;
+    }
+    case LogLevel.Warning: {
+      return theme.current.value.colors.warning;
+    }
+    case LogLevel.Error: {
+      return theme.current.value.colors.error;
+    }
+    case LogLevel.Critical: {
+      return theme.current.value.colors.secondary;
+    }
+    default: {
+      return theme.current.value.colors.primary;
```
I hadn't previously but I've now confirmed they do! It's a little unfortunate that the typings have `[key: string]: string` and but they extend the bases :(

aweebs commented on frontend/src/pages/settings/logs-and-activity.vue:
```diff
@@ -136,98 +132,128 @@ import IMdiStop from 'virtual:icons/mdi/stop';
-interface LoadingStatus {
-  status: 'loading' | 'loaded' | 'error';
-  errorMessage?: string;
+type LoadingStatus<T> =
+  | { status: 'loading' }
+  | { status: 'loaded'; data: T }
+  | { status: 'error'; message?: string };
```
I actually like this as a concept in certain cases, it's a more clear representation of what the state can be based on the status.
That being said, I agree that it's overkill and not useful here, since like you said we block the page while loading anyway. I'll rip it out :)

aweebs replied:
The one thing this does hide is the error state on the template, which is nice that it has. I'm a little split, it's easy to rip this all out, and if there's an error loading the logs, then just an empty list will be shown, which isn't the best UX.
On the other hand, we could just not await those initial data calls (or put them in an immediate only watcher), but the loading state is also just a blank list which isn't super clear for a user either and would require some more thought.

I'll get rid of it all for now along with translations for the error state in a separate commit just so it's easy to revert if we do want to keep it.

ferferga replied:
@aweebs I think the most reusable approach would be to always rely in Suspense for rendering components that are fast enough to load (not the case of library pages, for instance) and just show a generic snackbar if the load fails. Having each page handle their state like this is a nightmare for inconsistency.

The probability of errors in this component is really really low. Anyway, we have this somehow implemented with the Axios interceptors, but I have a local branch where I'm working in getting a progress bar at the top (in a similar fashion to YouTube), I can expand it for better global exception handling.

aweebs commented:
@ferferga thanks for going through so thoroughly!

I've made almost all of the changes and resolved the related comments that you left. Any unresolved comments are either because I have a follow up question or I've provided an explanation/change and would like you to re-review!

ThibaultNocchi commented:
I've got an error when leaving an item page (movie or episode) : `RequiredError: Required parameter itemId was null or undefined when calling getItem.`

aweebs commented:
> I've got an error when leaving an item page (movie or episode) : `RequiredError: Required parameter itemId was null or undefined when calling getItem.`

@ThibaultNocchi I'm not seeing the same using the deployment link above. Does it always happen or only certain items? Can you post a screenshot or the stack trace if you see it again?

ferferga commented:
@aweebs @ThibaultNocchi This commit removes the need to watch for route parameters to refresh the page: https://github.com/jellyfin/jellyfin-vue/pull/1899/commits/9d973d6dab71043c2af5138c509c1e981a4024a9
It doesn't affect query params, so search doesn't trigger a re-render, for instance

@aweebs Can you apply the changes of that commit to this branch (is a trivial change), test all the routes without watchers for path params and remove them if they work fine, please?
If they don't, report back please.

ferferga commented:
See [this](https://github.com/jellyfin/jellyfin-vue/pull/1899#issuecomment-1467255082) for blocking reasons.

aweebs commented:
> @aweebs @ThibaultNocchi This commit removes the need to watch for route parameters to refresh the page: [9d973d6](https://github.com/jellyfin/jellyfin-vue/commit/9d973d6dab71043c2af5138c509c1e981a4024a9) It doesn't affect query params, so search doesn't trigger a re-render, for instance
> 
> @aweebs Can you apply the changes of that commit to this branch (is a trivial change), test all the routes without watchers for path params and remove them if they work fine, please? If they don't, report back please.

I patched in the changes to `RouterViewTransition` and unfortunately the watchers were still necessary.

ferferga commented:
@aweebs This patch in the current state of this branch works perfectly for me:

```
diff --git a/frontend/src/components/RouterViewTransition.vue b/frontend/src/components/RouterViewTransition.vue
index 12d4902e..7f54f584 100644
--- a/frontend/src/components/RouterViewTransition.vue
+++ b/frontend/src/components/RouterViewTransition.vue
@@ -51,6 +51,6 @@ function getKey(route: RouteLocationNormalized): string {
     return route.meta.layout;
   }
 
-  return String(route.name);
+  return String(route.path);
 }
 </script>
diff --git a/frontend/src/pages/library/_itemId/index.vue b/frontend/src/pages/library/_itemId/index.vue
index 1c264bdb..53895306 100644
--- a/frontend/src/pages/library/_itemId/index.vue
+++ b/frontend/src/pages/library/_itemId/index.vue
@@ -55,7 +55,7 @@
 </template>
 
 <script setup lang="ts">
-import { computed, ref, watch } from 'vue';
+import { computed, ref, watch, onMounted } from 'vue';
 import { useI18n } from 'vue-i18n';
 import { useRoute } from 'vue-router';
 import { BaseItemDto, BaseItemKind } from '@jellyfin/sdk/lib/generated-client';
@@ -293,8 +293,8 @@ async function refreshItems(): Promise<void> {
   }
 }
 
-watch(() => (route.params as { itemId: string }).itemId, fetchLibrary, {
-  immediate: true
+onMounted(async () => {
+  await fetchLibrary(route.params.itemId as string);
 });
 
 watch(library, (lib) => {
```

aweebs commented:
@ferferga and @ThibaultNocchi I've replaced the route/query param watchers with the `onMounted` hook. This also fixed the issue with navigating away from an item page. I think the watcher was previously getting fired with an undefined itemId in the params.

The last 2 outstanding items for this PR are
1. https://github.com/jellyfin/jellyfin-vue/pull/1892#discussion_r1133168383 which I wasn't able to get around to tonight. I can do this as a follow up along with removing PlaybackInfoCard since they will both mean removing translations.
2. https://github.com/jellyfin/jellyfin-vue/pull/1892#discussion_r1133214361. This looks better than it did before but I'd like to make it a bit more robust for larger screen sizes by adding a max-width and some better breakpoints at smaller screen sizes.

ferferga commented:
@aweebs To be honest, we have a lot of old translations in the locale files, as people didn't care about it, so I wouldn't mind too much about them.

In fact I had planned to tackle the issue in my last commit before the merge or in the early days afterwards, since with Vue-i18n-extract that's much easier.

ferferga commented on frontend/src/components/Wizard/WizardMetadata.vue:
```diff
@@ -3,102 +3,108 @@
-    this.initialConfig = (
-      await getStartupApi(api).getStartupConfiguration()
-    ).data;
+const SomeItemSelectedRule = [(v: unknown[]): boolean | string => v.length > 0];
```
At this point I think we can extract these rules to something like ``utils/validation-rules.ts`` instead of duplicating them all over the place?

For now I think extracting this one is good, but we can look into doing the same thing for the rules used at the login pages.

ferferga commented on frontend/src/components/Wizard/WizardMetadata.vue:
```diff
@@ -3,102 +3,108 @@
-            PreferredMetadataLanguage: this.metadataCountry
-          }
-        });
+  initialConfig.value = (
+    await getStartupApi(api).getStartupConfiguration()
+  ).data;
```
I guess you've been playing this safe since we can't test these component properly because the missing Vuetify components, but I think it makes more sense to fetch all the required data at the page level and just leave the components show it. We've been repeating this request multiple times too.

ferferga replied:
My idea was passing the data as props from the page level, basically

aweebs replied:
I started to make this change, and realized that it will actually regress functionality. Going back to a previous step would not display the data that was previously saved anymore, since the config will only be fetched once at the page level.

eg. Go to Language step, set language and go to the next step. Then go back to the previous step, the language that is auto selected will not be what the user just chose.

The other option is to lift the load & update to the route level, but I think that's just moving code around for not much gain.

ferferga replied:
@aweebs Fair enough!

ferferga commented on frontend/src/pages/artist/_itemId/index.vue:
```diff
@@ -123,136 +119,120 @@ import {
-          userId: remote.auth.currentUserId
-        })
-      ).data.Items ?? [];
+onMounted(async () => {
```
Any reason why not an async function call at the top level?

aweebs replied:
I'm not sure that will work since the object return when `getItemById` is only reactive if it's found, and a straight undefined value if not. Going to leave this outside of the scope for this PR.

ferferga commented on frontend/src/pages/genre/_itemId/index.vue:
```diff
@@ -57,29 +60,54 @@ const items = itemsStore();
+const genre = ref<BaseItemDto>({});
+const genres = ref<BaseItemDto[]>([]);
+
+onMounted(async () => {
```
Same as above, I don't think we need the loading state in these components at all now. These components had this way of loading because they're SSR remnants

ferferga commented on frontend/src/pages/item/_itemId/index.vue:
```diff
@@ -246,92 +246,50 @@ import { getItemDetailsLink, getMediaStreams } from '@/utils/items';
+  crew.value.filter((person) => person.Type === 'Writer')
+);
+
+onMounted(async () => {
```
Same as above

ferferga commented on frontend/src/pages/library/_itemId/index.vue:
```diff
@@ -279,9 +293,7 @@ async function refreshItems(): Promise<void> {
-watch(() => (route.params as { viewId: string }).viewId, fetchLibrary, {
-  immediate: true
-});
+onMounted(() => fetchLibrary((route.params as { itemId: string }).itemId));
```
```suggestion
onMounted(async () => fetchLibrary((route.params as { itemId: string }).itemId));
```

FYI, this is the only place where is good to stick on onMounted because the server is too slow serving the request. With a better item store (which will result in a similar behaviour to the one we currently have in the userLibraries and index), this will probably be solved, but we still would need the request to be done here, so data is refreshed, but items ofc will be loaded much faster. Exactly the same thing that happens in index.

aweebs replied:
This only needs to be `async` if we need to `await` things, omitted for simplicity.

ferferga commented on frontend/src/pages/musicalbum/_itemId/index.vue:
```diff
@@ -75,45 +75,31 @@
-  methods: {
-    getItemDetailsLink
-  }
+onMounted(async () => {
```
Same as above

ferferga commented on frontend/src/pages/person/_itemId/index.vue:
```diff
@@ -125,98 +116,107 @@ import { getBlurhash } from '@/utils/images';
 
-route.meta.title = item.Name || '';
-route.meta.backdrop.blurhash = getBlurhash(item, ImageType.Backdrop);
+onMounted(async () => {
```
Same as above

ferferga commented on frontend/src/pages/series/_itemId/index.vue:
```diff
@@ -147,75 +147,42 @@ import { getBlurhash } from '@/utils/images';
+  crew.value.filter((person) => person.Type === 'Writer')
+);
+
+onMounted(async () => {
```
Same as above

ferferga commented on frontend/src/pages/wizard.vue:
```diff
@@ -78,82 +78,79 @@ meta:
+
+    await getStartupApi(api).completeWizard();
+    // Redirect to setup complete page
+    router.push('/server/login');
```
```suggestion
    router.replace('/server/login');
```

So user can't go back

aweebs commented on frontend/src/components/Item/Metadata/ImageSearch.vue:
```diff
@@ -3,59 +3,66 @@
           v-model="type"
           class="mx-4"
           :items="types"
+          item-title="text"
+          item-value="value"
```
Vuetify 3's default for the label is now `title` rather than `text` (https://vuetifyjs.com/en/api/v-select/). Leaving them blank results in
<img width="791" alt="image" src="https://user-images.githubusercontent.com/8809837/225546928-acb8888f-66a1-439d-9d34-99adead4b241.png">
:(

I had to make this change in a few uses of v-select to get them working again. This is something that could actually be enforced with typings on Vuetify's end to make sure that the default keys, or passed in keys, exist in the options object which are passed in and give a type error if not. Sadly I've learned not to expect much from Vuetify's types though.

I think there's 2 options here:
1. Update all places we pass in an options list to use the new default keys
2. Specify the text/value keys on the use of the component.

Because of the bad Vuetify typings, I think (2) is a bit more durable in the event of another Vuetify upgrade (however unlikely that may be) and my preference would be to leave this as is.

ferferga replied:
No problem, just leave this as is! I just wanted to make you aware of the defaults (at least according to their docs), but it's a super nitpicky thing that I don't care about!

aweebs commented on frontend/src/components/Item/Metadata/MetadataEditor.vue:
```diff
@@ -206,241 +216,230 @@ import { getUserLibraryApi } from '@jellyfin/sdk/lib/utils/api/user-library-api'
+    .value;
+});
+
+watch(() => props.itemId, getData, { immediate: true });
```
This doesn't cause a reference error and runs fine in practice, even with console logs to verify. Javascript scoping is cool with weird shit like that :(

I agree it reads better and makes more sense for uses to appear after declarations though, so I've moved the watch to the bottom of the file :)

aweebs commented on frontend/src/components/Item/Metadata/MetadataEditor.vue:
```diff
@@ -206,241 +216,230 @@ import { getUserLibraryApi } from '@jellyfin/sdk/lib/utils/api/user-library-api'
+
+const dateCreated = computed(() => {
+  if (!metadata.value?.DateCreated) {
+    return '';
```
This gets passed directly into the `v-date` component which expects a `string` for the value prop, it can't be undefined :(

aweebs commented on frontend/src/components/Item/Metadata/MetadataEditor.vue:
```diff
@@ -206,241 +216,230 @@ import { getUserLibraryApi } from '@jellyfin/sdk/lib/utils/api/user-library-api'
-  width: 100%;
+<style lang="scss" scoped>
+.person-icon {
+  background-color: rgb(var(--v-secondary-darken1));
```
Removed this class from this file.

However I couldn't find any documentation on Vuetify's css variables that it provides. I used my browser to see what was actually defined, and it doesn't look like theme lighten classes exist for `pages/settings/metadata.vue`.
<img width="403" alt="image" src="https://user-images.githubusercontent.com/8809837/225555614-29fcf5e9-87e6-467c-a556-4fd34e6edea0.png">
 I'll remove that style for now.

For `pages/settings/index.vue`, the `--v-theme-card` is correctly defined (as are the rest of the uses of `--v-theme-*` css variables in other components) and removing it does make a difference on the UI so I will leave that one alone.

ferferga commented:
@aweebs Force pushed to include the latest Vuetify hotfix (fixes were very relevant for vlistitem and vtabs, very used in this client). I also took the liberty to remove [this unused import](https://github.com/jellyfin/jellyfin-vue/pull/1892/files#diff-f15af6c8ecda3e9f0735cf68727f4bbc44c51014ef618ea355e6e85555228262R111) and squash it with the onMounted commit (the one that last modified that line). Hopefully it's a minor annoyance for you.

I didn't had the time to properly test the reactivity of the item store, but as they're getters, I believe they're fully reactive when invoked normally (they are in the playbackManager store, just not when returned by the Axios interceptors). That's why I think it's worth a shot, but don't be worried about ``undefined``. You can also check playbackManager store, we have a lot of examples that can return undefined and still they're reactive. Getters seem to be like computed properties (don't know exactly why, though, but I guess this was the foundation for the old Vue 2 reactivity system based on getters/setters).

ferferga reviewed (approved):
I tried a bit with the commented reactivity bits for the pages (as they were the last concern I had) but the changes required go way beyond the scope of this PR.

Thank you very much for all your work!

ThibaultNocchi reviewed (approved):
LGTM, I think ferf will squash a few commits
